File: zte-agent/src/wifi.rs
use std::process::Command;

use serde_json::{json, Value};

use crate::handlers::AppState;
use crate::ubus;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn uci_get_wireless(key: &str) -> String {
    ubus::uci_get(&format!("wireless.{key}")).unwrap_or_default()
}

fn uci_get_mbb(key: &str) -> String {
    ubus::uci_get(&format!("zte_mbb.{key}")).unwrap_or_default()
}

fn iw_info(iface: &str) -> (String, String) {
    let output = Command::new("iw")
        .args([iface, "info"])
        .output()
        .ok();
    let out = output
        .map(|o| String::from_utf8_lossy(&o.stdout).to_string())
        .unwrap_or_default();
    let channel = out
        .lines()
        .find_map(|l| {
            let l = l.trim();
            if l.starts_with("channel ") {
                l.split_whitespace().nth(1).map(|s| s.to_string())
            } else {
                None
            }
        })
        .unwrap_or_default();
    let bw = out
        .lines()
        .find_map(|l| {
            let l = l.trim();
            if let Some(pos) = l.find("width:") {
                let rest = l[pos + 6..].trim();
                let end = rest.find("MHz").map(|i| i + 3).unwrap_or(rest.len());
                Some(rest[..end].trim().to_string())
            } else {
                None
            }
        })
        .unwrap_or_default();
    (channel, bw)
}

fn station_count(iface: &str) -> u64 {
    let output = Command::new("sh")
        .args(["-c", &format!("iw {iface} station dump 2>/dev/null | grep -c Station")])
        .output()
        .ok();
    output
        .and_then(|o| {
            String::from_utf8_lossy(&o.stdout)
                .trim()
                .parse::<u64>()
                .ok()
        })
        .unwrap_or(0)
}

fn sanitize_uci_value(v: &str) -> String {
    v.chars()
        .filter(|c| !matches!(c, '\'' | '"' | ';' | '$' | '`' | '\\' | '|' | '<' | '>' | '&'))
        .collect()
}

// ---------------------------------------------------------------------------
// GET /api/wifi/status
// ---------------------------------------------------------------------------

pub fn wifi_status(_state: &AppState) -> (u16, Value) {
    let mut result = serde_json::Map::new();

    // Global switches from zte_mbb
    let mut wifi_onoff = uci_get_mbb("wifi.wifi_onoff");
    let mut wifi6_switch = uci_get_mbb("wifi.wifi6_switch");

    // Fallback: if mbb UCI keys are missing, read from zwrt_wlan report
    if wifi_onoff.is_empty() || wifi6_switch.is_empty() {
        if let Ok(report) = ubus::call("zwrt_wlan", "report", Some("{}")) {
            if wifi_onoff.is_empty() {
                wifi_onoff = report["wifi_onoff"]
                    .as_str()
                    .unwrap_or("1")
                    .to_string();
            }
            if wifi6_switch.is_empty() {
                wifi6_switch = report["wifi6_switch"]
                    .as_str()
                    .unwrap_or("0")
                    .to_string();
            }
        }
    }
    result.insert("wifi_onoff".into(), json!(wifi_onoff));
    result.insert("wifi6_switch".into(), json!(wifi6_switch));

    // Radio config
    result.insert("radio2_disabled".into(), json!(uci_get_wireless("wifi0.disabled")));
    result.insert("radio5_disabled".into(), json!(uci_get_wireless("wifi1.disabled")));
    result.insert("channel_2g".into(), json!(uci_get_wireless("wifi0.channel")));
    result.insert("channel_5g".into(), json!(uci_get_wireless("wifi1.channel")));
    result.insert("txpower_2g".into(), json!(uci_get_wireless("wifi0.txpowerpercent")));
    result.insert("txpower_5g".into(), json!(uci_get_wireless("wifi1.txpowerpercent")));
    result.insert("htmode_2g".into(), json!(uci_get_wireless("wifi0.htmode")));
    result.insert("htmode_5g".into(), json!(uci_get_wireless("wifi1.htmode")));
    result.insert("country_code".into(), json!(uci_get_wireless("wifi0.country")));

    // Interface config
    result.insert("ssid_2g".into(), json!(uci_get_wireless("main_2g.ssid")));
    result.insert("ssid_5g".into(), json!(uci_get_wireless("main_5g.ssid")));
    result.insert("key_2g".into(), json!(uci_get_wireless("main_2g.key")));
    result.insert("key_5g".into(), json!(uci_get_wireless("main_5g.key")));
    result.insert("encryption_2g".into(), json!(uci_get_wireless("main_2g.encryption")));
    result.insert("encryption_5g".into(), json!(uci_get_wireless("main_5g.encryption")));
    result.insert("hidden_2g".into(), json!(uci_get_wireless("main_2g.hidden")));
    result.insert("hidden_5g".into(), json!(uci_get_wireless("main_5g.hidden")));

    // Runtime info from iw
    let (ch2, bw2) = iw_info("wlan0");
    let (ch5, bw5) = iw_info("wlan2");
    result.insert("actual_channel_2g".into(), json!(ch2));
    result.insert("actual_bw_2g".into(), json!(bw2));
    result.insert("actual_channel_5g".into(), json!(ch5));
    result.insert("actual_bw_5g".into(), json!(bw5));

    // Client counts
    let c2g = station_count("wlan0");
    let c5g = station_count("wlan2");
    result.insert("clients_2g".into(), json!(c2g));
    result.insert("clients_5g".into(), json!(c5g));
    result.insert("clients_total".into(), json!(c2g + c5g));

    // Guest WiFi summary
    result.insert("guest_disabled_2g".into(), json!(uci_get_wireless("guest_2g.disabled")));
    result.insert("guest_disabled_5g".into(), json!(uci_get_wireless("guest_5g.disabled")));
    result.insert("guest_ssid".into(), json!(uci_get_wireless("guest_2g.ssid")));

    (200, json!({"ok": true, "data": result}))
}

// ---------------------------------------------------------------------------
// PUT /api/wifi/settings
// ---------------------------------------------------------------------------

pub fn wifi_set(_state: &AppState, body: &[u8]) -> (u16, Value) {
    let parsed: Value = match serde_json::from_slice(body) {
        Ok(v) => v,
        Err(_) => return (400, json!({"ok": false, "error": "invalid JSON"})),
    };
    let obj = match parsed.as_object() {
        Some(o) => o,
        None => return (400, json!({"ok": false, "error": "expected JSON object"})),
    };
    // Serialise against every other writer of the `wireless` package (the
    // scenario engine, u60-guard, homemode). Held until the reload has been
    // verified, not just until commit — see wifi_radio's lock notes.
    let lk = match crate::wifi_radio::lock(crate::wifi_radio::HTTP_WAIT) {
        Ok(l) => l,
        Err(e) => return (503, json!({"ok": false, "error": e})),
    };

    let uci_map: &[(&str, &str)] = &[
        ("ssid_2g", "wireless.main_2g.ssid"),
        ("ssid_5g", "wireless.main_5g.ssid"),
        ("key_2g", "wireless.main_2g.key"),
        ("key_5g", "wireless.main_5g.key"),
        ("encryption_2g", "wireless.main_2g.encryption"),
        ("encryption_5g", "wireless.main_5g.encryption"),
        ("hidden_2g", "wireless.main_2g.hidden"),
        ("hidden_5g", "wireless.main_5g.hidden"),
        ("channel_2g", "wireless.wifi0.channel"),
        ("channel_5g", "wireless.wifi1.channel"),
        ("txpower_2g", "wireless.wifi0.txpowerpercent"),
        ("txpower_5g", "wireless.wifi1.txpowerpercent"),
        ("htmode_2g", "wireless.wifi0.htmode"),
        ("htmode_5g", "wireless.wifi1.htmode"),
        ("radio2_disabled", "wireless.wifi0.disabled"),
        ("radio5_disabled", "wireless.wifi1.disabled"),
    ];
    let mbb_map: &[(&str, &str)] = &[
        ("wifi_onoff", "zte_mbb.wifi.wifi_onoff"),
        ("wifi6_switch", "zte_mbb.wifi.wifi6_switch"),
    ];

    let txpower_keys: &[&str] = &["txpower_2g", "txpower_5g"];

    let mut wireless_changed = false;
    let mut mbb_changed = false;
    let mut only_txpower = true;
    let mut txpower_2g_val: Option<u32> = None;
    let mut txpower_5g_val: Option<u32> = None;

    // Region/country code applies to BOTH radios (wifi0 + wifi1). It is the
    // self-managed phy's regulatory domain, so it forces a full wlan reload
    // (re-insmod) — never a txpower hot-apply.
    if let Some(country) = obj.get("country").and_then(|v| v.as_str()) {
        let cc = sanitize_uci_value(country).to_uppercase();
        if cc.len() == 2 && cc.chars().all(|c| c.is_ascii_alphabetic()) {
            for path in ["wireless.wifi0.country", "wireless.wifi1.country"] {
                let current = ubus::uci_get(path).unwrap_or_default();
                if current != cc {
                    if let Err(e) = ubus::uci_set_no_commit(path, &cc) {
                        return (500, json!({"ok": false, "error": e}));
                    }
                    wireless_changed = true;
                    only_txpower = false;
                }
            }
        }
    }

    for (key, value) in obj {
        let val_str = match value {
            Value::String(s) => s.clone(),
            Value::Number(n) => n.to_string(),
            Value::Bool(b) => if *b { "1" } else { "0" }.to_string(),
            _ => continue,
        };
        let val_str = sanitize_uci_value(&val_str);

        // Check wireless UCI map
        if let Some(&(_, path)) = uci_map.iter().find(|&&(k, _)| k == key) {
            let current = ubus::uci_get(path).unwrap_or_default();
            if current != val_str {
                if let Err(e) = ubus::uci_set_no_commit(path, &val_str) {
                    return (500, json!({"ok": false, "error": e}));
                }
                wireless_changed = true;
                if !txpower_keys.contains(&key.as_str()) {
                    only_txpower = false;
                } else if key == "txpower_2g" {
                    txpower_2g_val = val_str.parse().ok();
                } else if key == "txpower_5g" {
                    txpower_5g_val = val_str.parse().ok();
                }
            }
            continue;
        }

        // Check mbb map
        if let Some(&(_, path)) = mbb_map.iter().find(|&&(k, _)| k == key) {
            let current = ubus::uci_get(path).unwrap_or_default();
            if current != val_str {
                // mbb keys are firmware-dependent; skip silently if missing
                if ubus::uci_set_no_commit(path, &val_str).is_ok() {
                    mbb_changed = true;
                    only_txpower = false;
                }
            }
        }
    }

    // Commit batched changes
    if wireless_changed {
        if let Err(e) = ubus::uci_commit("wireless") {
            return (500, json!({"ok": false, "error": e}));
        }
    }
    if mbb_changed {
        if let Err(e) = ubus::uci_commit("zte_mbb") {
            return (500, json!({"ok": false, "error": e}));
        }
    }

    if !wireless_changed && !mbb_changed {
        return (200, json!({"ok": true, "data": {"status": "ok", "note": "no changes"}}));
    }

    // Hot-apply txpower if that's the only change
    if wireless_changed && only_txpower {
        if let Some(val) = txpower_2g_val {
            let _ = Command::new("iw")
                .args(["dev", "wlan0", "set", "txpower", "limit", &(val * 30).to_string()])
                .output();
        }
        if let Some(val) = txpower_5g_val {
            let _ = Command::new("iw")
                .args(["dev", "wlan2", "set", "txpower", "limit", &(val * 30).to_string()])
                .output();
        }
        return (200, json!({"ok": true, "data": {"status": "ok", "hot": true}}));
    }

    // Full reload needed. It runs on its own thread, which keeps the lock until
    // the reload is verified; the response does not wait for it.
    if wireless_changed {
        crate::wifi_radio::finish_in_background(lk, "settings");
    }

    (200, json!({"ok": true, "data": {"status": "ok"}}))
}

// ---------------------------------------------------------------------------
// GET /api/wifi/guest
// ---------------------------------------------------------------------------

pub fn guest_status(_state: &AppState) -> (u16, Value) {
    let mut result = serde_json::Map::new();

    result.insert("ssid".into(), json!(uci_get_wireless("guest_2g.ssid")));
    result.insert("key".into(), json!(uci_get_wireless("guest_2g.key")));
    result.insert("encryption".into(), json!(uci_get_wireless("guest_2g.encryption")));
    result.insert("disabled_2g".into(), json!(uci_get_wireless("guest_2g.disabled")));
    result.insert("disabled_5g".into(), json!(uci_get_wireless("guest_5g.disabled")));
    result.insert("hidden".into(), json!(uci_get_wireless("guest_2g.hidden")));
    result.insert("isolate".into(), json!(uci_get_wireless("guest_2g.isolate")));
    result.insert("guest_active_time".into(), json!(uci_get_wireless("guest_2g.guest_active_time")));

    // Runtime remaining time
    let remaining = ubus::call("zwrt_wlan", "wlan_get_guest_access_left_time", Some("{}"))
        .ok()
        .and_then(|v| v["guest_left_time"].as_str().and_then(|s| s.parse::<i64>().ok()))
        .unwrap_or(-1);
    result.insert("remaining_seconds".into(), json!(remaining));

    (200, json!({"ok": true, "data": result}))
}

// ---------------------------------------------------------------------------
// PUT /api/wifi/guest
// ---------------------------------------------------------------------------

pub fn guest_set(_state: &AppState, body: &[u8]) -> (u16, Value) {
    let parsed: Value = match serde_json::from_slice(body) {
        Ok(v) => v,
        Err(_) => return (400, json!({"ok": false, "error": "invalid JSON"})),
    };
    let obj = match parsed.as_object() {
        Some(o) => o,
        None => return (400, json!({"ok": false, "error": "expected JSON object"})),
    };

    let guest_map: &[(&str, &[&str])] = &[
        ("guest_ssid", &["wireless.guest_2g.ssid", "wireless.guest_5g.ssid"]),
        ("guest_key", &["wireless.guest_2g.key", "wireless.guest_5g.key"]),
        ("guest_encryption", &["wireless.guest_2g.encryption", "wireless.guest_5g.encryption"]),
        ("guest_disabled", &["wireless.guest_2g.disabled", "wireless.guest_5g.disabled"]),
        ("guest_disabled_2g", &["wireless.guest_2g.disabled"]),
        ("guest_disabled_5g", &["wireless.guest_5g.disabled"]),
        ("guest_hidden", &["wireless.guest_2g.hidden", "wireless.guest_5g.hidden"]),
        ("guest_isolate", &["wireless.guest_2g.isolate", "wireless.guest_5g.isolate"]),
        ("guest_active_time", &["wireless.guest_2g.guest_active_time", "wireless.guest_5g.guest_active_time"]),
    ];

    let lk = match crate::wifi_radio::lock(crate::wifi_radio::HTTP_WAIT) {
        Ok(l) => l,
        Err(e) => return (503, json!({"ok": false, "error": e})),
    };

    let mut changed = false;
    for (key, value) in obj {
        let val_str = match value {
            Value::String(s) => s.clone(),
            Value::Number(n) => n.to_string(),
            Value::Bool(b) => if *b { "1" } else { "0" }.to_string(),
            _ => continue,
        };
        let val_str = sanitize_uci_value(&val_str);

        if let Some(&(_, paths)) = guest_map.iter().find(|&&(k, _)| k == key) {
            for &path in paths {
                // Guest interfaces may not exist; skip silently if missing
                if ubus::uci_set_no_commit(path, &val_str).is_ok() {
                    changed = true;
                }
            }
        }
    }

    if !changed {
        return (200, json!({"ok": true, "data": {"status": "ok", "note": "no changes"}}));
    }

    // Commit batched changes
    if let Err(e) = ubus::uci_commit("wireless") {
        return (500, json!({"ok": false, "error": e}));
    }

    crate::wifi_radio::finish_in_background(lk, "guest");

    (200, json!({"ok": true, "data": {"status": "ok"}}))
}

File: web/src/lib/api/schemas/wifi.ts
// Response shapes for the Wi-Fi, guest Wi-Fi, AP radio and Home Mode endpoints
// of zte-agent (zte-agent/src/wifi.rs, wifi_radio.rs, homemode.rs).
//
// Every handler here builds its JSON itself, so the Rust shape is the source
// of truth; page-side disagreements are noted inline. Types only — no runtime
// code.

/**
 * GET /api/wifi/status — wifi.rs:80 `wifi_status`.
 *
 * Every string field is a raw `uci get` value; a missing key (or a failed
 * `uci get`) is "" — never absent, never null (wifi.rs:12-18).
 * `wifi_onoff` / `wifi6_switch` come from `zte_mbb.wifi.*`, falling back to
 * `ubus zwrt_wlan report`, then to "1" / "0" (wifi.rs:84-103).
 *
 * Page disagreements:
 * - router/wifi/page.tsx types this as `Record<string, string>`, but the three
 *   `clients_*` fields are numbers.
 * - router/wifi/page.tsx also falls back to `ssid`, `password`, `key`,
 *   `encryption`, `channel`, `bandwidth`, `tx_power` — the handler never
 *   returns any of those names.
 */
export interface WifiStatus {
  /** "1" = on, "0" = off (zte_mbb.wifi.wifi_onoff). */
  wifi_onoff: string;
  /** "1" = 802.11ax on (zte_mbb.wifi.wifi6_switch). */
  wifi6_switch: string;
  /** wireless.wifi0.disabled — "1" = 2.4 GHz radio off; "" = enabled. */
  radio2_disabled: string;
  /** wireless.wifi1.disabled — "1" = 5 GHz radio off. */
  radio5_disabled: string;
  /** Configured channel ("auto" or a number as a string). */
  channel_2g: string;
  channel_5g: string;
  /** wireless.wifiN.txpowerpercent: "25" | "50" | "100" on the page. */
  txpower_2g: string;
  txpower_5g: string;
  /** wireless.wifiN.htmode, e.g. "HT20", "VHT80", "auto". */
  htmode_2g: string;
  htmode_5g: string;
  /** wireless.wifi0.country (two letters, upper case). Written via PUT `country`. */
  country_code: string;
  ssid_2g: string;
  ssid_5g: string;
  /** Plain-text passphrases. */
  key_2g: string;
  key_5g: string;
  /** "psk2+ccmp" | "psk3+ccmp" | "psk2+psk3+ccmp" | "none". */
  encryption_2g: string;
  encryption_5g: string;
  /** "1" = hidden. */
  hidden_2g: string;
  hidden_5g: string;
  /** From `iw wlan0 info` ("channel N ..."); "" when the interface is down. */
  actual_channel_2g: string;
  /** From `iw wlan0 info` "width: 20 MHz" → "20 MHz"; "" when down. */
  actual_bw_2g: string;
  /** From `iw wlan2 info`. */
  actual_channel_5g: string;
  actual_bw_5g: string;
  /** `iw wlanN station dump | grep -c Station`; 0 on failure. */
  clients_2g: number;
  clients_5g: number;
  clients_total: number;
  /** wireless.guest_2g.disabled / guest_5g.disabled; "" if the section is absent. */
  guest_disabled_2g: string;
  guest_disabled_5g: string;
  /** wireless.guest_2g.ssid. */
  guest_ssid: string;
}

/**
 * PUT /api/wifi/settings body — wifi.rs:155 `wifi_set`.
 * Any subset; values may be strings, numbers or booleans (true→"1").
 * Unknown keys are ignored. `country` must be two letters or it is ignored.
 */
export interface WifiSettingsBody {
  ssid_2g?: string;
  ssid_5g?: string;
  key_2g?: string;
  key_5g?: string;
  encryption_2g?: string;
  encryption_5g?: string;
  hidden_2g?: string;
  hidden_5g?: string;
  channel_2g?: string;
  channel_5g?: string;
  txpower_2g?: string;
  txpower_5g?: string;
  htmode_2g?: string;
  htmode_5g?: string;
  radio2_disabled?: string;
  radio5_disabled?: string;
  wifi_onoff?: string;
  wifi6_switch?: string;
  country?: string;
}

/**
 * PUT /api/wifi/settings response data — wifi.rs:276/291/300.
 * `note: "no changes"` when nothing differed; `hot: true` when only txpower
 * changed (applied with `iw set txpower`); otherwise just `status` and the
 * reload runs in the background after the reply.
 */
export interface WifiSettingsResult {
  status: "ok";
  note?: "no changes";
  hot?: true;
}

/**
 * GET /api/wifi/radio — wifi_radio.rs:329 `radio_get` (→ `observe`, :268).
 * Read-only: uci get + `ps w` + `hostapd_cli status` per AP interface.
 */
export interface WifiRadio {
  /** AP interfaces as configured in uci (main_2g/main_5g.disabled != "1"). Not the radios. */
  configured: { ap_2g: boolean; ap_5g: boolean };
  /** Running hostapd processes; usize::MAX (18446744073709551615) if `ps` failed. */
  hostapd_processes: number;
  /** hostapd_cli reports state=ENABLED on that interface. */
  beaconing: { wlan0: boolean; wlan2: boolean };
}

/** PUT /api/wifi/radio body — wifi_radio.rs:339. Missing key = keep current. */
export interface WifiRadioSetBody {
  ap_2g?: boolean;
  ap_5g?: boolean;
}

/**
 * PUT /api/wifi/radio response data — wifi_radio.rs:310-317. Only returned
 * when verified; otherwise 503 with the observed state in the error string.
 */
export interface WifiRadioSetResult extends WifiRadio {
  verified: boolean;
  target: { ap_2g: boolean; ap_5g: boolean };
  /** Present only when `ubus zwrt_wlan reload` returned an error. */
  reload_error?: string;
}

/**
 * GET /api/wifi/guest — wifi.rs:307 `guest_status`.
 * All strings are raw `uci get wireless.guest_2g.*` ("" when absent).
 *
 * Page (router/wifi-guest/page.tsx) also declares `guest_onoff`, `guest_ssid`,
 * `guest_key`, `guest_encryption`, `guest_isolate`, `active_time`,
 * `guest_disabled_2g`, `guest_disabled_5g` — the handler returns none of
 * those names (they are the PUT body names); the page falls back correctly.
 */
export interface GuestWifi {
  ssid: string;
  key: string;
  encryption: string;
  /** "1" = guest 2.4 GHz AP disabled. */
  disabled_2g: string;
  disabled_5g: string;
  hidden: string;
  /** "1" = client isolation. */
  isolate: string;
  /** Minutes, "0" = unlimited (unit from the page label; unconfirmed). */
  guest_active_time: string;
  /**
   * `ubus zwrt_wlan wlan_get_guest_access_left_time` → guest_left_time, parsed
   * to an integer; -1 when the call fails. Seconds (page counts it down per
   * second). What the firmware returns while the guest AP is off is
   * unconfirmed.
   */
  remaining_seconds: number;
}

/** PUT /api/wifi/guest body — wifi.rs:333. Each key writes guest_2g and/or guest_5g. */
export interface GuestWifiBody {
  guest_ssid?: string;
  guest_key?: string;
  guest_encryption?: string;
  /** Sets both bands. */
  guest_disabled?: string;
  guest_disabled_2g?: string;
  guest_disabled_5g?: string;
  guest_hidden?: string;
  guest_isolate?: string;
  guest_active_time?: string;
}

/** PUT /api/wifi/guest response data — wifi.rs:381/391. */
export interface GuestWifiResult {
  status: "ok";
  note?: "no changes";
}

/**
 * GET /api/homemode — homemode.rs:103 `homemode_get`. Reads local files only.
 * Also the response of PUT /api/homemode (homemode.rs:192).
 */
export interface HomeMode {
  /** `/data/homemode/disabled` does not exist. */
  enabled: boolean;
  /** Configured list, or the built-in defaults when using_default. */
  ssids: string[];
  /** `/data/homemode/ssids` absent (an empty file is an explicit empty list). */
  using_default: boolean;
  default_ssids: string[];
  /** First word of `/data/homemode/state`; "normal" when absent. */
  mode: "home" | "normal";
  /** mode === "home" (the worker has turned Wi-Fi off). */
  wifi_off: boolean;
  /** Minutes between rechecks while at home, 1–60 (default 2). */
  check_every: number;
  /** Missed rechecks before Wi-Fi returns, 1–30 (default 2). */
  exit_misses: number;
}

/** PUT /api/homemode body — homemode.rs:132. Numbers may also be numeric strings. */
export interface HomeModeBody {
  enabled?: boolean;
  ssids?: string[];
  check_every?: number | string;
  exit_misses?: number | string;
}

/**
 * GET /api/homemode/scan — homemode.rs:235 `homemode_scan`.
 * One entry per SSID (strongest BSS kept), strongest first; hidden SSIDs
 * dropped. `signal` is dBm (f64, e.g. -48).
 *
 * page expects an optional `note` (router/home-mode/page.tsx:132); the
 * handler never returns it, and returns `woke_radio`, which the page ignores.
 */
export interface HomeModeScan {
  networks: { ssid: string; signal: number }[];
  /** 2.4 GHz was off (home mode) and was woken just for this scan. */
  woke_radio: boolean;
}

/**
 * GET /api/homemode/log — homemode.rs:314 `homemode_log`.
 * Last 200 lines of each file, oldest first, "\n"-joined; "" when absent.
 * Line format (scripts/homemode.sh:99): `<date +%Y-%m-%dT%H:%M:%S%z> <message>`.
 */
export interface HomeModeLog {
  /** /data/log/homemode.log — Wi-Fi on/off transitions. */
  events: string;
  /** /data/log/homemode-scan.log — periodic rechecks while at home. */
  scans: string;
}

/** GET endpoints of this area (path without query → response data). */
export interface WifiGetMap {
  "/api/wifi/status": WifiStatus;
  "/api/wifi/radio": WifiRadio;
  "/api/wifi/guest": GuestWifi;
  "/api/homemode": HomeMode;
  "/api/homemode/scan": HomeModeScan;
  "/api/homemode/log": HomeModeLog;
}
